lib/prx_access/prx_hyper_resource.rb

```diff
@@ -0,0 +1,141 @@
+require 'hyperresource'
+require 'active_support/core_ext/hash'
+
+module PRXAccess
+
+  class PRXHyperResource < HyperResource
+    def incoming_body_filter(hash)
+      super(hash.deep_transform_keys { |key| key.to_s.underscore })
+    end
+
+    def outgoing_body_filter(hash)
+      super(hash.deep_transform_keys { |key| key.to_s.camelize(:lower) })
+    end
+
+    class Link < HyperResource::Link
+      attr_accessor :type, :profile
+
+      def initialize(resource, link_spec={})
+        super
+        self.type = link_spec['type']
+        self.profile = link_spec['profile']
+      end
+
+      def where(params)
+        super.tap do |res|
+          res.type = self.type
+          res.profile = self.profile
+        end
+      end
+
+      def headers(*args)
+        super.tap do |res|
+          if args.count > 0
+            res.type = self.type
+            res.profile = self.profile
+          end
+        end
+      end
+
+      def post_response(attrs=nil)
+        attrs ||= self.resource.attributes
+        attrs = (self.resource.default_attributes || {}).merge(attrs)
+
+        # adding this line to call outgoing_body_filter
+        attrs = resource.outgoing_body_filter(attrs)
+
+        response = faraday_connection.post do |req|
+          req.body = self.resource.adapter.serialize(attrs)
+        end
+        response
+      end
+
+      def put_response(attrs=nil)
+        attrs ||= self.resource.attributes
+        attrs = (self.resource.default_attributes || {}).merge(attrs)
+
+        # adding this line to call outgoing_body_filter
+        attrs = resource.outgoing_body_filter(attrs)
+
+        response = faraday_connection.put do |req|
+          req.body = self.resource.adapter.serialize(attrs)
+        end
+        response
+      end
+
+      def patch_response(attrs=nil)
+        attrs ||= self.resource.attributes.changed_attributes
+        attrs = (self.resource.default_attributes || {}).merge(attrs)
+
+        # adding this line to call outgoing_body_filter
+        attrs = resource.outgoing_body_filter(attrs)
+
+        response = faraday_connection.patch do |req|
+          req.body = self.resource.adapter.serialize(attrs)
+        end
+        response
+      end
+    end
+  end
+
+  def default_headers
+    {
+      'Content-Type' => 'application/json',
+      'Accept' =>  'application/json'
+    }
+  end
+
+  def api(options = {})
```

Comment:
Sent this over slack to peter, could we also allow passing in an auth header?
Could be a PR after this I guess, but it's a 1 liner:
```ruby
  def api(options = {})
    opts = { root: cms_root, headers: default_headers }.merge(options)
    if !opts[:headers]['Authorization'] && account = opts.delete(:account)  # <-- check here for blank header
      token = get_account_token(account)
      opts[:headers]['Authorization'] = "Bearer #{token}"
    end

    PRXHyperResource.new(opts)
  end
```